sklearn/tree/_criterion.pxd

```diff
@@ -52,10 +52,24 @@ cdef class Criterion:
                                 double* impurity_right) nogil
     cdef void node_value(self, double* dest) nogil
     cdef double impurity_improvement(self, double impurity_parent,
-                                     double impurity_left,
-                                     double impurity_right) nogil
+                                    double impurity_left,
+                                    double impurity_right) nogil
     cdef double proxy_impurity_improvement(self) nogil
 
+cdef class Criterion(BaseCriterion):
+    # The supervised criterion computes the impurity of a node and the reduction of
+    # impurity of a split on that node using the distribution of labels in parent and
+    # children nodes. It also computes the output statistics
+    # such as the mean in regression and class probabilities in classification.
+
+    # Internal structures
+    cdef const DOUBLE_t[:, ::1] y        # Values of y
+
+    # Methods
+    cdef int init(self, const DOUBLE_t[:, ::1] y, DOUBLE_t* sample_weight,
+                  double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                  SIZE_t end) nogil except -1
```

Comment:
Also, it might be possible to use `const DOUBLE_t[:]` for `sample_weights`.

Reply:
Making a note here for a future issue:

To explore changing this to a memview instead of a pointer, see https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html.

What you will need to do is change it here in the function header, in the function instantiation and then ctrl+f anything that "calls" sample_weight. However, imo after ctrl+fing in the tree/ submodule, this would be a larger change that I would be in favor of doing in a separate PR given PRing into scikit-learn is already pretty challenging.

Original comment: https://github.com/scikit-learn/scikit-learn/pull/24678/files#r1013113564

Reply:
Made an issue.